README.md

```diff
@@ -174,95 +174,151 @@ For pyenv to install python correctly you should [**install the Python build dep
-     ~~~
-
-   - For **Zsh**:
-     ~~~ zsh
-     echo -e 'if command -v pyenv 1>/dev/null 2>&1; then\n  eval "$(pyenv init -)"\nfi' >> ~/.zshrc
-     ~~~
-
-   - For **Fish shell**:
-     Add this to `~/.config/fish/config.fish`:
-     ~~~ fish
-     pyenv init - | source
-     ~~~
-
-    **General warning**: There are some systems where the `BASH_ENV` variable is configured
-    to point to `.bashrc`. On such systems you should almost certainly put the above mentioned line
-    `eval "$(pyenv init -)"` into `.bash_profile`, and **not** into `.bashrc`. Otherwise you
-    may observe strange behaviour, such as `pyenv` getting into an infinite loop.
-    See [#264](https://github.com/pyenv/pyenv/issues/264) for details.
-
-4. **Restart your shell so the path changes take effect.**
-   You can now begin using pyenv.
-    ```sh
-    exec "$SHELL"
-    ```
+2. **Configure your shell's environment for Pyenv**
+
+   **Note:** The below instructions for specific shells are designed for common shell setups.  
+   If you have an uncommon setup and they don't work for you,
+   use the guidance text and the [Advanced Configuration](#advanced-configuration)
+   section below to figure out what you need to do in your specific case.
+   
+   **MacOS note:** If you installed Pyenv with Homebrew, you don't need
+   to add the `PYENV_ROOT=` and `PATH=` lines.
+   
+   1. **Adjust the session-wide environment for your account.** Define
+   the `PYENV_ROOT` environment variable to point to the path where
+   you cloned the Pyenv repo, add the `pyenv` command-line utility to your `PATH`,
+   run the output of `pyenv init --path` to enable shims.
+   
+      These commands need to be added into your shell startup files in such a way
+      that _they are executed only once per session, by its login shell._
+      This typically means they need to be added into a per-user shell-specific
+      `~/.*profile` file, _and_ into `~/.profile`, too, so that they are also
+      run by GUI managers (which typically act as a `sh` login shell).
+
+      - For **bash**:
+
+         ~~~ bash
+         echo 'export PYENV_ROOT="$HOME/.pyenv"' >> ~/.profile
+         echo 'export PATH="$PYENV_ROOT/bin:$PATH"' >> ~/.profile
+         echo 'eval "$(pyenv init --path)"' >> ~/.profile
+         ~~~
+
+         - **If your `~/.profile` sources `~/.bashrc` (Debian, Ubuntu, Mint):**
+
+            Put these lines into `~/.profile` _before_ the part that sources `~/.bashrc`:
+            ~~~bash
+            export PYENV_ROOT="$HOME/.pyenv"
+            export PATH="$PYENV_ROOT/bin:$PATH"
+            eval "$(pyenv init --path)"
+            ~~~
+
+            <!--This is an alternative option and needn't be replicated to `pyenv init`-->
+            Alternatively, for an automated installation, you can run the following:
+            ~~~ bash
+            echo -e 'if shopt -q login_shell; then' \
```

Comment:
@native-api I'm wondering if it might be easier / more robust to just check for either a guard variable e.g. `[ -z "$PYENV_ROOT" ]` or test for `pyenv` with `! command -v pyenv &> /dev/null` instead of making assumptions about *why* a particular RC file has been sourced?

Reply:
That's not an "assumption", that's actually a requirement :-) See the guidance text for this section in the README when this code needs to run.

> if it might be easier / more robust to just check

There's no such foolproof test. E.g. `command -v pyenv` will always succeed if `pyenv` is installed into a location that's already in `PATH` (e.g. with Homebrew).

The solution to #1649 that we chose is to conform to the UNIX design: make shell configuration commands "dumb" and rely on the shell configuration infrastructure to run them at the right moments.
The only alternative was to second-guess the user -- and thus inevitably guess wrong from time to time.

Reply:
Actually, if you have a problem understanding how this chunk achieves the goal stated in the guidance text -- maybe I should add a brief explanation text?

Reply:
@native-api no, I'm not confused; I didn't see that you scoped this section to the distros that source bashrc from profile. With that context, I think this is as general + safe as we can be here.